# Univariate Analysis Case

## Background

In financial companies, we often encounter an upward trend in the first overdue monthly loan contracts rate. The first overdue rate generally refers to the number of customers who have not repaid on time when only the first period is due on a specific repayment date. The monitoring and effective reduction of the first overdue rate are essential to a company's asset quality analysis. 

Here in this dataset, daily monitoring has discovered that a certain consumer loan product's first overdue rate is gradually increasing. And we need to lower the first overdue rate to reduce the loss caused by the product.

Considering the protection of customers' privacy, no further specific information about this product would be exposed in this case. 

## Goal:
Through data exploration and analysis, develop a strategy that can effectively reduce the first overdue rate.

## Analysis Process
1. **Access Data**: Because the strategy we want to analyze will be used to determine whether the customer will be overdue when the customer applies, the basic idea of the strategy analysis is to restore the data of these customers with first overperformance at the time of application (this restoration refers to extracting cusotmer information of each dimension at the time of application, the more the better), and then use these data to find variables that can distinguish good customers from bad customers and formulate strategies.

2. **Filter Out Valid Variables**: We will use the method of univariate analysis. In this case, we will group the number of ***credit inquiries times*** and ***credit rating levels*** to see the impact on overdue rate. The main purpose of using univariate analysis is to rule out variables with obvious distinction between good and bad in order to formulate our strategy.

3. **Calculate the Lifting Degree**: The Lifting Degree = The proportion of worst overdue group to the total overdue customers/ The Proportion of customers in this group

4. **Develop the Strategy**: Based on the lift degree formulated in the previous step, simulate the decline in the first overdue rate after the strategy is executed.

### 1. Access Data

In [16]:
import pandas as pd
import numpy as np

import pandas as pd
import sqlalchemy

In [17]:
# Read data
engine = sqlalchemy.create_engine('mysql+pymysql://frogdata05:Frogdata!1321@localhost:3306/froghd')

sql_cmd = "select * from load_datas"

dt = pd.read_sql(sql=sql_cmd, con=engine)

In [18]:
dt.head()

user_id  age occupation  work_times  credit_score credit_level  \
0        1   22        个体户        45.0         650.0            A   
1        2   28         职员        34.0         680.0            A   
2        3   39         其他       113.0         530.0            A   
3        4   24         职员        44.0         800.0            A   
4        5   30         职员        82.0           NaN            A   

  credit_check_times  credit_card_use_rate  is_overdue  
0                 23                  0.00           0  
1                 33                  0.32           0  
2                 32                  0.05           0  
3                 35                  0.01           0  
4                 28                  0.00           0

In [19]:
#Check the overall dataset
dt.shape

(56456, 9)

In [20]:
# Calculate the Overall Fisrt Overdue Rate = The Amount of Overdue Rate Customers / The Amount of Overall Applicants
 #"o" means overdue, "1" means not overdue
dt["is_overdue"].sum()/dt.shape[0]

0.307584667705824

According to the result, the overall first overdue rate is generally high at around 30.8%

### 2. Filter Out Valid Variables

In financial companies' daily work, there will be teams specifically responsible for crawling, calculating, and processing variable data. They are continually acquiring and processing data that may help risk control and provide them to the risk control team. And risk control team would need to explore from these thousands of variables that can handle the risk of overdue but at the same time not mistakenly reject many customers with good credit situation.


#### 2.1 Group by the ***Credit Inquiry Times***

In [21]:
# Define a function of grouping by credit inquiry times
# Generally, the higher times of one's credit inquiry is, the higher the one's overdue risk would be
# Segment the credit inquiry times into 6 groups

def judge_zhengxin(nums):
    try:
        nums = int(nums)
        if 0<=nums and nums<3:
            return '1:[0,3)'
        elif 3<=nums and nums<6:
            return '2:[3,6)'
        elif 6<=nums and nums<12:
            return '3:[6,12)'
        elif 12<=nums and nums<21:
            return '4:[12,21)'
        elif 21<=nums:
            return '5:[21,infinite)'
    except Exception as e:
        return "6:Missing"

In [22]:
# Create a new field to record grouping information
dt['credit_inquiry_group']=dt['credit_check_times'].apply(judge_zhengxin)

In [23]:
dt.head()

user_id  age occupation  work_times  credit_score credit_level  \
0        1   22        个体户        45.0         650.0            A   
1        2   28         职员        34.0         680.0            A   
2        3   39         其他       113.0         530.0            A   
3        4   24         职员        44.0         800.0            A   
4        5   30         职员        82.0           NaN            A   

  credit_check_times  credit_card_use_rate  is_overdue credit_inquiry_group  
0                 23                  0.00           0      5:[21,infinite)  
1                 33                  0.32           0      5:[21,infinite)  
2                 32                  0.05           0      5:[21,infinite)  
3                 35                  0.01           0      5:[21,infinite)  
4                 28                  0.00           0      5:[21,infinite)

In [24]:
# Summarize the statistics of each group
# The proportion of interval users and the number of customers not overdue shall be calculated separately

dt_info = dt.groupby("credit_inquiry_group").agg({
    "user_id":'count',
    "is_overdue":sum
}).reset_index().rename(columns={"user_id":"Number of Interval Customers",
                                "is_overdue":"Number of Interval Overdue Customers",
                                "credit_inquiry_group" : "Credit Inquiry Group"})
dt_info

Credit Inquiry Group  Number of Interval Customers  \
0              1:[0,3)                         14667   
1              2:[3,6)                         15469   
2             3:[6,12)                         15316   
3            4:[12,21)                          6577   
4      5:[21,infinite)                          3213   
5            6:Missing                          1214   

   Number of Interval Overdue Customers  
0                                  3352  
1                                  4157  
2                                  4786  
3                                  2739  
4                                  1923  
5                                   408

In [25]:
# Further, we calculate the proportion of interval users and the number of customers not overdue 
# Then we get the result of the first overdue rate within intervals

dt_info['The Proportion of Interval Users']=dt_info["Number of Interval Customers"]/dt_info["Number of Interval Customers"].sum()
dt_info["The Number of Non-overdue Users"]=dt_info["Number of Interval Customers"]-dt_info["Number of Interval Overdue Customers"]
dt_info["The Interval First Overdue Rate"]=dt_info["Number of Interval Overdue Customers"]/dt_info["Number of Interval Customers"]
dt_info

Credit Inquiry Group  Number of Interval Customers  \
0              1:[0,3)                         14667   
1              2:[3,6)                         15469   
2             3:[6,12)                         15316   
3            4:[12,21)                          6577   
4      5:[21,infinite)                          3213   
5            6:Missing                          1214   

   Number of Interval Overdue Customers  The Proportion of Interval Users  \
0                                  3352                          0.259795   
1                                  4157                          0.274001   
2                                  4786                          0.271291   
3                                  2739                          0.116498   
4                                  1923                          0.056912   
5                                   408                          0.021503   

   The Number of Non-overdue Users  The Interval First Overdue Rate  
0                            11315                         0.228540  
1                            11312                         0.268731  
2                            10530                         0.312484  
3                             3838                         0.416451  
4                             1290                         0.598506  
5                              806                         0.336079

#### 2.2 Group by the ***Credit Rating Levels***

In [26]:
# Define a function of credit rating levels, segement them into 5 groups

def judge_pingji(level):
    if level=="A":
        return "A"
    elif level=="AA":
        return "AA"
    elif level in ("B","C","D"):
        return "BCD"
    elif level in ("E","HR","NC"):
        return "ERC"
    else:
        return "Missing"


In [27]:
# Create a new field to record grouping information

dt['Credit Level Group']=dt['credit_level'].apply(judge_pingji)

In [28]:
dt.head()

user_id  age occupation  work_times  credit_score credit_level  \
0        1   22        个体户        45.0         650.0            A   
1        2   28         职员        34.0         680.0            A   
2        3   39         其他       113.0         530.0            A   
3        4   24         职员        44.0         800.0            A   
4        5   30         职员        82.0           NaN            A   

  credit_check_times  credit_card_use_rate  is_overdue credit_inquiry_group  \
0                 23                  0.00           0      5:[21,infinite)   
1                 33                  0.32           0      5:[21,infinite)   
2                 32                  0.05           0      5:[21,infinite)   
3                 35                  0.01           0      5:[21,infinite)   
4                 28                  0.00           0      5:[21,infinite)   

  Credit Level Group  
0                  A  
1                  A  
2                  A  
3                  A  
4                  A

In [29]:
# Summarize the statistics of each group
# The proportion of interval users and the number of customers not overdue shall be calculated separately

dt_info2 = dt.groupby("Credit Level Group").agg({
    "user_id":'count',
    "is_overdue":sum
}).reset_index().rename(columns={"user_id":"Number of Interval Customers",
                                "is_overdue":"Number of Interval Overdue Customers"})
dt_info2

Credit Level Group  Number of Interval Customers  \
0                  A                          3369   
1                 AA                          3784   
2                BCD                         16320   
3                ERC                          8060   
4            Missing                         24923   

   Number of Interval Overdue Customers  
0                                   851  
1                                   640  
2                                  5920  
3                                  4251  
4                                  5703

In [30]:
# Further, we calculate the proportion of interval users and the number of customers not overdue 
# Then we get the result of the first overdue rate within intervals


dt_info2["The Proportion of Interval Users"]=dt_info2["Number of Interval Customers"]/dt_info2["Number of Interval Customers"].sum()
dt_info2["The Number of Non-overdue Users"]=dt_info2["Number of Interval Customers"]-dt_info2["Number of Interval Overdue Customers"]
dt_info2["The Interval First Overdue Rate"]=dt_info2["Number of Interval Overdue Customers"]/dt_info2["Number of Interval Customers"]
dt_info2

Credit Level Group  Number of Interval Customers  \
0                  A                          3369   
1                 AA                          3784   
2                BCD                         16320   
3                ERC                          8060   
4            Missing                         24923   

   Number of Interval Overdue Customers  The Proportion of Interval Users  \
0                                   851                          0.059675   
1                                   640                          0.067026   
2                                  5920                          0.289075   
3                                  4251                          0.142766   
4                                  5703                          0.441459   

   The Number of Non-overdue Users  The Interval First Overdue Rate  
0                             2518                         0.252597  
1                             3144                         0.169133  
2                            10400                         0.362745  
3                             3809                         0.527419  
4                            19220                         0.228825

In [31]:
# The first overdue rate in different credit rating level groups
dt_info

Credit Inquiry Group  Number of Interval Customers  \
0              1:[0,3)                         14667   
1              2:[3,6)                         15469   
2             3:[6,12)                         15316   
3            4:[12,21)                          6577   
4      5:[21,infinite)                          3213   
5            6:Missing                          1214   

   Number of Interval Overdue Customers  The Proportion of Interval Users  \
0                                  3352                          0.259795   
1                                  4157                          0.274001   
2                                  4786                          0.271291   
3                                  2739                          0.116498   
4                                  1923                          0.056912   
5                                   408                          0.021503   

   The Number of Non-overdue Users  The Interval First Overdue Rate  
0                            11315                         0.228540  
1                            11312                         0.268731  
2                            10530                         0.312484  
3                             3838                         0.416451  
4                             1290                         0.598506  
5                              806                         0.336079

### 3. Calculate the Lifting Degree

Generally speaking, it measures overall risk control's improvement effect after rejecting the worst part of customers.
The higher the lifting degree is, it means that this variable can more effectively distinguish between good and bad customers, and less falsely reject good customers.
 
Formula:  The Lift Degree = The proportion of worst overdue group to the total overdue customers/ The Proportion of customers in this group

This lifting degree index is generally used for relative comparison in a batch of variable analysis. In many cases, the variable with higher lift is selected as a strategy from the limited variables.

**As shown below, by sorting the lifting degree of all variables in descending order, we found that variables credit inquiry times and credit rating levels have the highest lift degrees, reaching around 1.94 and 1.71.**


In [40]:
# Check on our credit inquiry times group first
dt_info

Credit Inquiry Group  Number of Interval Customers  \
0              1:[0,3)                         14667   
1              2:[3,6)                         15469   
2             3:[6,12)                         15316   
3            4:[12,21)                          6577   
4      5:[21,infinite)                          3213   
5            6:Missing                          1214   

   Number of Interval Overdue Customers  The Proportion of Interval Users  \
0                                  3352                          0.259795   
1                                  4157                          0.274001   
2                                  4786                          0.271291   
3                                  2739                          0.116498   
4                                  1923                          0.056912   
5                                   408                          0.021503   

   The Number of Non-overdue Users  The Interval First Overdue Rate  
0                            11315                         0.228540  
1                            11312                         0.268731  
2                            10530                         0.312484  
3                             3838                         0.416451  
4                             1290                         0.598506  
5                              806                         0.336079

In [38]:
# Calculate the lifting degree of each indicator - credit inquiry times
# The Lifting Degree = The proportion of worst overdue group to the total overdue customers/ The Proportion of customers in this group
# The worst group here is the group with the highest first overdue rate after grouped by credit information

dt_info["The Interval First Overdue Rate"].idxmax()
# dt_info.iloc[dt_info["The Interval First Overdue Rate"].idxmax()]
bad_rate = dt_info.iloc[dt_info["The Interval First Overdue Rate"].idxmax()]["Number of Interval Overdue Customers"]/dt_info["Number of Interval Overdue Customers"].sum()
num_rate = dt_info.iloc[dt_info["The Interval First Overdue Rate"].idxmax()]["Number of Interval Customers"]/dt_info["Number of Interval Customers"].sum()
bad_rate/num_rate

1.9458254325820934

In [41]:
# Now check on our credit level groups
dt_info2

Credit Level Group  Number of Interval Customers  \
0                  A                          3369   
1                 AA                          3784   
2                BCD                         16320   
3                ERC                          8060   
4            Missing                         24923   

   Number of Interval Overdue Customers  The Proportion of Interval Users  \
0                                   851                          0.059675   
1                                   640                          0.067026   
2                                  5920                          0.289075   
3                                  4251                          0.142766   
4                                  5703                          0.441459   

   The Number of Non-overdue Users  The Interval First Overdue Rate  
0                             2518                         0.252597  
1                             3144                         0.169133  
2                            10400                         0.362745  
3                             3809                         0.527419  
4                            19220                         0.228825

In [39]:
# # Calculate the lifting degree of each indicator - credit rating levels
# dt_info2.iloc[dt_info2["The Interval First Overdue Rate"].idxmax()]

bad_rate2 = dt_info2.iloc[dt_info2["The Interval First Overdue Rate"].idxmax()]["Number of Interval Overdue Customers"]/dt_info2["Number of Interval Overdue Customers"].sum()
num_rate2 = dt_info2.iloc[dt_info2["The Interval First Overdue Rate"].idxmax()]["Number of Interval Customers"]/dt_info2["Number of Interval Customers"].sum()
bad_rate2/num_rate2

1.7147127611157038

### 4. Develop the Strategy

Through the univariate analysis in previous steps, we have filtered out the two variables with the highest lift degree, namely the variable of ***credit inquiry times*** and ***the credit rating levels***. Now let's take a look at the impact on the overall overdue if the worst group customers of these two variables are precluded from our dataset. 

This impact refers to the decrease in the first overdue rate of the remaining customers compared to the one before the exclusion, assuming that we rule out all the 3213 customers whose credit inquiry times >= 21.

In [36]:

new_overdue_nums = dt_info["Number of Interval Overdue Customers"].sum() - dt_info.iloc[dt_info["The Interval First Overdue Rate"].idxmax()]["Number of Interval Overdue Customers"]
new_overdue_rate = new_yuqi_nums / (dt_info2["Number of Interval Customers"].sum() - dt_info.iloc[dt_info["The Interval First Overdue Rate"].idxmax()]["Number of Interval Customers"])

old_overdue_nums = dt_info["Number of Interval Overdue Customers"].sum()
old_overdue_rate = old_yuqi_nums / dt_info["Number of Interval Customers"].sum()

new_overdue_rate

0.2900287361718911

In [37]:
# Check on the previous overdue rate
old_yuqi_rate

0.307584667705824

In [24]:
# The new overdue rate has dropped by 1.76%
old_yuqi_rate-new_yuqi_rate

0.017555931533932867

Clearly, by ruling out all the 3213 customers whose credit inquiry times >=21, othe first overdue rate of our product is declined by around 1.76%, indicating that our strategy worked. 